src/server/cluster/incoming_slot_migration.cc

```diff
@@ -21,46 +22,62 @@ using absl::GetFlag;
 // It is created per shard on the target node to initiate FLOW step.
 class ClusterShardMigration {
  public:
-  ClusterShardMigration(uint32_t shard_id, Service* service) : source_shard_id_(shard_id) {
-    executor_ = std::make_unique<JournalExecutor>(service);
+  ClusterShardMigration(uint32_t shard_id, Service* service)
+      : source_shard_id_(shard_id), socket_(nullptr), executor_(service) {
   }
 
-  void Start(Context* cntx, util::FiberSocketBase* source) {
-    socket_ = source;
+  void Start(Context* cntx, util::FiberSocketBase* source, util::fb2::BlockingCounter bc) {
+    {
+      std::lock_guard lk(mu_);
+      socket_ = source;
+    }
+
+    absl::Cleanup cleanup([this]() {
+      std::lock_guard lk(mu_);
+      socket_ = nullptr;
+    });
     JournalReader reader{source, 0};
     TransactionReader tx_reader{false};
 
     while (!cntx->IsCancelled()) {
-      if (cntx->IsCancelled())
-        break;
-
       auto tx_data = tx_reader.NextTxData(&reader, cntx);
       if (!tx_data) {
+        // TODO add error processing
         VLOG(1) << "No tx data";
         break;
       }
 
-      if (tx_data->opcode == journal::Op::FIN) {
-        VLOG(2) << "Flow " << source_shard_id_ << " is finalized";
-        break;
-      } else if (tx_data->opcode == journal::Op::PING) {
+      while (tx_data->opcode == journal::Op::FIN) {
+        VLOG(2) << "Attemot to finalize flow " << source_shard_id_;
+        bc->Dec();  // we can Join the flow now
+        // if we get new data attempt is failed
+        if (tx_data = tx_reader.NextTxData(&reader, cntx); !tx_data) {
+          VLOG(1) << "Correct finish for flow " << source_shard_id_;
+          return;
+        }
+        bc->Add();  // the flow isn't finihed so we lock it again
```

Comment:
Isn't it a problem that we first Dec(), making this joinable, and then Add()? What happens if someone attempts to join in between?

Reply:
it's OK if we get this situation, the error happens so we can lose some data. It's according design